pkg/scheduler/api/hyper_node_info.go

```diff
@@ -0,0 +1,518 @@
+		if hnInfo == nil || hnInfo.HyperNode == nil {
+			continue
+		}
+
+		isLeaf := true
+		hasRegexMatch := false
+		for _, member := range hnInfo.HyperNode.Spec.Members {
+			if member.Type == topologyv1alpha1.MemberTypeHyperNode {
+				isLeaf = false
+				break
+			}
+			if member.Selector.RegexMatch != nil {
+				hasRegexMatch = true
+			}
+		}
+		if isLeaf && hasRegexMatch {
+			leaf.Insert(name)
+		}
+	}
+	return leaf
+}
+
+// setParent sets the parent of a HyperNode member.
+func (hni *HyperNodesInfo) setParent(member, parent string) error {
+	hn, ok := hni.hyperNodes[member]
+	if !ok {
+		klog.InfoS("HyperNode not exists in cache, maybe not created or not be watched, will set parent first", "name", member, "parent", parent)
+		hn = NewHyperNodeInfo(&topologyv1alpha1.HyperNode{ObjectMeta: metav1.ObjectMeta{
+			Name: member,
+		}})
+		hni.hyperNodes[member] = hn
+		hn.parent = parent
+		return nil
+	}
+	currentParent := hn.parent
+	if currentParent == "" {
+		hn.parent = parent
+		return nil
+	}
+
+	if currentParent != parent {
+		return fmt.Errorf("HyperNode %s already has a parent %s, and cannot set another parent %s", member, currentParent, parent)
+	}
+
+	return nil
+}
+
+// getMembers retrieves the members of a HyperNode based on the selector.
+func (hni *HyperNodesInfo) getMembers(selector topologyv1alpha1.MemberSelector, nodes []*corev1.Node) sets.Set[string] {
+	if selector.ExactMatch != nil {
+		if selector.ExactMatch.Name == "" {
+			return sets.New[string]()
+		}
+		return sets.New[string](selector.ExactMatch.Name)
+	}
+
+	members := sets.New[string]()
+	if selector.RegexMatch != nil {
+		pattern := selector.RegexMatch.Pattern
+		reg, err := regexp.Compile(pattern)
+		if err != nil {
+			klog.ErrorS(err, "Failed to compile regular expression", "pattern", pattern)
+			return sets.Set[string]{}
+		}
+		for _, node := range nodes {
+			if reg.MatchString(node.Name) {
+				members.Insert(node.Name)
+			}
+		}
+	}
+	return members
+}
+
+// exactMatchMember retrieves the member name if the selector is an exact match.
+func (hni *HyperNodesInfo) exactMatchMember(selector topologyv1alpha1.MemberSelector) string {
+	if selector.ExactMatch != nil {
+		return selector.ExactMatch.Name
+	}
+	return ""
+}
+
+func (hni *HyperNodesInfo) updateAncestors(name string) error {
+	ancestors := hni.GetAncestors(name)
+	// Clear realNodesSet of hyperNode before rebuild hyperNode cache.
+	for ancestor := range ancestors {
+		delete(hni.realNodesSet, ancestor)
+		hni.resetParent(ancestor)
+	}
+
+	processed := sets.New[string]()
+	ancestorsChain := sets.New[string]()
+	nodes, err := hni.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Failed to list nodes", "hyperNodeName", name)
+		return err
+	}
+
+	for ancestor := range ancestors {
+		if hn, ok := hni.hyperNodes[ancestor]; ok {
+			if err := hni.BuildHyperNodeCache(hn, processed, ancestorsChain, nodes); err != nil {
+				hni.setReady(false)
```

Comment:
Would it be better if we moved `hni.setReady(false)` to the beginning of updateAncestors? If a job is submitted when rebuilding the HyperNodes tree, it may be scheduled to the wrong node at this time.